### DISCUSSION TITLE
How can I overriding the default logger (with logrus or pick your prefered logging library)

### DISCUSSION THREAD

**Discussion by bsergean** [anacrolix/torrent#551] [Q&A] [answered]
Hi there,

It looks like this commit https://github.com/anacrolix/torrent/issues/271 allowed users to override the client logging.

Here is an example, from the cmd/torrent. -> https://github.com/anacrolix/torrent/blob/master/cmd/torrent/main.go#L321
```
    if flags.Quiet {
        clientConfig.Logger = log.Discard
    }
```

This Discard object is a global thing, defined as: https://github.com/anacrolix/log/blob/master/global.go#L14

```
var (
    Default = Logger{StreamLogger{
        W:   os.Stderr,
        Fmt: LineFormatter,
    }}
    Discard = Logger{StreamLogger{
        W:   ioutil.Discard,
        Fmt: func(Msg) []byte { return nil },
    }}
)
```

A stream logger is this:

```
type StreamLogger struct {
    W   io.Writer
    Fmt ByteFormatter
}

func (me StreamLogger) Log(msg Msg) {
    me.W.Write(me.Fmt(msg.Skip(1)))
}
```

My application is using logrus, and I would like to capture logs and emit them with logrus. So I guess I just need to implement a Log function that takes a msg and print it (with logrus ?). I'll try that.

```
func (me StreamLogger) Log(msg Msg) {
    me.W.Write(me.Fmt(msg.Skip(1)))
}
```

**Comment by bsergean** [comment, marked as answer]
I figured it out, here's what I did:

Calling the thing:

```
package mypackage_name
...
clientConfig := torrent.NewDefaultClientConfig()
clientConfig.Logger = GetTorrentLogger()
```

Override logic

```
// Override torrent log library with logrus
// See https://github.com/anacrolix/torrent/discussions/551

package mypackage_name

import (
	torrentlog "github.com/anacrolix/log"
	log "github.com/sirupsen/logrus"
)

type LogRusAdapterLogger struct{}

func (l LogRusAdapterLogger) Log(msg torrentlog.Msg) {
	if level, ok := msg.GetLevel(); ok {
		// A "DEBUG" is available, we ignore it as it's very noisy, but we should expose it
		switch level.LogString() {
		case "INFO":
			log.Info("⛲ " + msg.Text())
		case "WARN":
			log.Warn("⛲ " + msg.Text())
		case "ERROR":
			log.Error("⛲ " + msg.Text())
		case "CRIT":
			log.Error("⛲ " + msg.Text()) // no logrus.Critical, report this as an error
		case "FATAL":
			log.Error("⛲ " + msg.Text()) // not sure what to do, also map to an error
		}
	}
}

func GetTorrentLogger() torrentlog.Logger {
	return torrentlog.Logger{LogRusAdapterLogger{}}
}
```

**Reply by anacrolix** [reply]
You might match against the level values rather than their String, for example `log.Level`. Additionally the silly package name convention for `anacrolix/log` is `analog`, if you care to use that. :)

**Comment by anacrolix** [comment]
Thanks for contributing that. What do you think of the interface? I'm not especially happy with anacrolix/log, it "bubbles" up log messages in a Context-like chain, but it means that filtering doesn't occur for some time, and there's a fair bit of allocation prior to hitting a filter. Additionally the filters can exist anywhere in the chain, rather than in a single place. It's very flexible, but not practical I think.

**Comment by bsergean** [comment]
I haven’t thought too much about the interface, as long as I can capture the logs it’s fine for me at this point.

I’ll use analog it sounds better.
I tried using log.Level (the ‘int’) but I was getting some odd errors while the ’string interface worked.

**Comment by skinner12** [comment]
I've tried but I get this error:

`./main.go:39:21: msg.GetLevel undefined (type "github.com/anacrolix/log".Msg has no field or method GetLevel)
./main.go:57:46: implicit assignment of unexported field 'loggerCore' in "github.com/anacrolix/log".Logger literal
./main.go:57:46: cannot use LogRusAdapterLogger{} (type LogRusAdapterLogger) as type "github.com/anacrolix/log".loggerCore in field value`

**Reply by anacrolix** [reply]
Sorry this will have changed. There was an optimization to check levels without traversing a chain of message context, part of that is that Level is no longer part of the Msg. If it's needed to adapt to other loggers, I could make some changes to pass that through to the formatter.
